pkg/query-service/telemetry/telemetry.go

```diff
@@ -227,141 +223,171 @@ func createTelemetry() {
-					if err != nil && getSamplesInfoInLastHeartBeatInterval == 0 {
-						telemetry.activeUser["metrics"] = 0
-					}
-				}
-				if (telemetry.activeUser["traces"] != 0) || (telemetry.activeUser["metrics"] != 0) || (telemetry.activeUser["logs"] != 0) {
-					telemetry.activeUser["any"] = 1
-				}
-				telemetry.SendEvent(TELEMETRY_EVENT_ACTIVE_USER, map[string]interface{}{
-					"traces":  telemetry.activeUser["traces"],
-					"metrics": telemetry.activeUser["metrics"],
-					"logs":    telemetry.activeUser["logs"],
-					"any":     telemetry.activeUser["any"]},
-					"", true, false)
-				telemetry.activeUser = map[string]int8{"traces": 0, "metrics": 0, "logs": 0, "any": 0}
 
-			case <-ticker.C:
+	// Define heartbeat function
+	heartbeatFunc := func() {
+		tagsInfo, _ := telemetry.reader.GetTagsInfoInLastHeartBeatInterval(context.Background(), HEART_BEAT_DURATION)
 
-				tagsInfo, _ := telemetry.reader.GetTagsInfoInLastHeartBeatInterval(context.Background(), HEART_BEAT_DURATION)
+		if len(tagsInfo.Env) != 0 {
+			telemetry.SendEvent(TELEMETRY_EVENT_ENVIRONMENT, map[string]interface{}{"value": tagsInfo.Env}, "", true, false)
+		}
 
-				if len(tagsInfo.Env) != 0 {
-					telemetry.SendEvent(TELEMETRY_EVENT_ENVIRONMENT, map[string]interface{}{"value": tagsInfo.Env}, "", true, false)
-				}
+		languages := []string{}
+		for language := range tagsInfo.Languages {
+			languages = append(languages, language)
+		}
+		if len(languages) > 0 {
+			telemetry.SendEvent(TELEMETRY_EVENT_LANGUAGE, map[string]interface{}{"language": languages}, "", true, false)
+		}
+		services := []string{}
+		for service := range tagsInfo.Services {
+			services = append(services, service)
+		}
+		if len(services) > 0 {
+			telemetry.SendEvent(TELEMETRY_EVENT_SERVICE, map[string]interface{}{"serviceName": services}, "", true, false)
+		}
+		totalSpans, _ := telemetry.reader.GetTotalSpans(context.Background())
+		totalLogs, _ := telemetry.reader.GetTotalLogs(context.Background())
+		spansInLastHeartBeatInterval, _ := telemetry.reader.GetSpansInLastHeartBeatInterval(context.Background(), HEART_BEAT_DURATION)
+		getSamplesInfoInLastHeartBeatInterval, _ := telemetry.reader.GetSamplesInfoInLastHeartBeatInterval(context.Background(), HEART_BEAT_DURATION)
+		totalSamples, _ := telemetry.reader.GetTotalSamples(context.Background())
+		tsInfo, _ := telemetry.reader.GetTimeSeriesInfo(context.Background())
+
+		getLogsInfoInLastHeartBeatInterval, _ := telemetry.reader.GetLogsInfoInLastHeartBeatInterval(context.Background(), HEART_BEAT_DURATION)
+
+		traceTTL, _ := telemetry.reader.GetTTL(context.Background(), &model.GetTTLParams{Type: constants.TraceTTL})
+		metricsTTL, _ := telemetry.reader.GetTTL(context.Background(), &model.GetTTLParams{Type: constants.MetricsTTL})
+		logsTTL, _ := telemetry.reader.GetTTL(context.Background(), &model.GetTTLParams{Type: constants.LogsTTL})
```

Comment:
Context usage doesn't look right.

We should create single context and pass it here.

Reply:
Done